File: bindings/node/examples/sym_self.js
'use strict';
const { ffi, memory: mem } = require('../');

console.log('urb-ffi ffi.sym_self demo');

// strlen e puts estão no processo sem dlopen
const strlenDesc = ffi.describe('u64 strlen(cstring)');
const putsDesc = ffi.describe('i32 puts(cstring)');
const getenvDesc = ffi.describe('cstring getenv(cstring)');
const strlen = ffi.bind(ffi.sym_self('strlen'), strlenDesc);
const puts   = ffi.bind(ffi.sym_self('puts'), putsDesc);

const msg = 'hello world from sym_self';
puts(msg);
console.log(`strlen("${msg}") =`, strlen(msg));

const path = ffi.bind(ffi.sym_self('getenv'), getenvDesc)('PATH');
console.log('PATH (primeiros 60 chars):', path ? path.slice(0, 60) + '...' : '(nulo)');

File: bindings/lua/examples/sym_self.lua
local source = debug.getinfo(1, "S").source:sub(2)
local dir = source:match("^(.*)/[^/]+$") or "."
package.path = dir .. "/../?.lua;" .. package.path

local urb = require("urb_ffi")
local ffi = urb.ffi

print("urb-ffi ffi.sym_self demo (lua)")

local puts_ptr = ffi.sym_self("puts")
local strlen_ptr = ffi.sym_self("strlen")
local puts_desc = ffi.describe("i32 puts(cstring)")
local strlen_desc = ffi.describe("u64 strlen(cstring)")
local puts = ffi.bind(puts_ptr, puts_desc)
local strlen = ffi.bind(strlen_ptr, strlen_desc)
local text = "hello world from sym_self"

puts(text)
print(string.format("strlen(%q) = %s", text, tostring(strlen(text))))
